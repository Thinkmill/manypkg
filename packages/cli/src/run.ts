import { getPackages } from "@manypkg/get-packages";
import path from "path";
import spawn from "spawndamnit";
import * as logger from "./logger";
import { ExitError } from "./errors";

export async function runCmd(args: string[], cwd: string) {
  let { packages, root } = await getPackages(cwd);

  const exactMatchingPackage = packages.find(pkg => {
    return (
      pkg.packageJson.name === args[0] ||
      path.relative(root.dir, pkg.dir) === args[0]
    );
  });

  if (exactMatchingPackage) {
    const { code } = await spawn("yarn", args.slice(1), {
      cwd: exactMatchingPackage.dir,
      stdio: "inherit"
    });
    throw new ExitError(code);
  }

  const matchingPackages = packages.filter(pkg => {
    return (
      pkg.packageJson.name.includes(args[0]) ||
      path.relative(root.dir, pkg.dir).includes(args[0])
    );
  });

  if (matchingPackages.length > 1) {
    logger.error(
      `an identifier must only match a single package but "${
        args[0]
      }" matches the following packages: \n${matchingPackages
        .map(x => x.packageJson.name)
        .join("\n")}`
    );
    throw new ExitError(1);
  } else if (matchingPackages.length === 0) {
    logger.error("No matching packages found");
    throw new ExitError(1);
  } else {
    const { code } = await spawn("yarn", args.slice(1), {
      cwd: matchingPackages[0].dir,
      stdio: "inherit"
    });
    throw new ExitError(code);
  }
}
