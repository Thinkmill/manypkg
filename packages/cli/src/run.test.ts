import { runCmd } from "./run";
import fixturez from "fixturez";
import spawn from "spawndamnit";
import stripAnsi from "strip-ansi";

const f = fixturez(__dirname);

describe("Run command", () => {
  it.each([
    ["package-one", "start", 0],
    ["package-one", "test", 0],
    ["packages/package-two", "start", 0],
    ["package-two-one", "start", 0],
    ["package", "start", 1],
    ["package-two", "start", 1],
    ["package-two-one", "something", 1],
    ["package-three", "start", 1],
    ["pkg-one", "start", 0],
    ["pkg-two", "start", 1],
    ["@manypkg/basic-fixture-pkg-two", "start", 0],
    ["pkg-two-one", "start", 0]
    // @ts-ignore
  ])(
    'should execute "%s %s" and exit with %i',
    async (arg0, arg1, exitCode) => {
      const { stdout, stderr, code } = await spawn(
        require.resolve("../bin"),
        // @ts-ignore
        ["run", arg0, arg1],
        {
          cwd: f.find("basic-with-scripts")
        }
      );
      expect(code).toBe(exitCode);
      expect(stripAnsi(stdout.toString())).toMatchSnapshot("stdout");
      expect(stripAnsi(stderr.toString())).toMatchSnapshot("stderr");
    }
  );
});
