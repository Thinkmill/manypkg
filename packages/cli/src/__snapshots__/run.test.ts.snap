// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Run command should execute "@manypkg/basic-fixture-pkg-two start" and exit with 0: stderr 1`] = `
"Browserslist: caniuse-lite is outdated. Please run next command \`yarn upgrade\`
warning package.json: No license field
warning ../../package.json: No license field
"
`;

exports[`Run command should execute "@manypkg/basic-fixture-pkg-two start" and exit with 0: stdout 1`] = `
"$ echo hello start
hello start
"
`;

exports[`Run command should execute "package start" and exit with 1: stderr 1`] = `
"Browserslist: caniuse-lite is outdated. Please run next command \`yarn upgrade\`
☔️ error an identifier must only match a single package but \\"package\\" matches the following packages: 
☔️ error @manypkg/basic-fixture-pkg-one
☔️ error @manypkg/basic-fixture-pkg-two
☔️ error @manypkg/basic-fixture-pkg-two-one
"
`;

exports[`Run command should execute "package start" and exit with 1: stdout 1`] = `""`;

exports[`Run command should execute "package-one start" and exit with 0: stderr 1`] = `
"Browserslist: caniuse-lite is outdated. Please run next command \`yarn upgrade\`
warning package.json: No license field
warning ../../package.json: No license field
"
`;

exports[`Run command should execute "package-one start" and exit with 0: stdout 1`] = `
"$ echo hello start
hello start
"
`;

exports[`Run command should execute "package-one test" and exit with 0: stderr 1`] = `
"Browserslist: caniuse-lite is outdated. Please run next command \`yarn upgrade\`
warning package.json: No license field
warning ../../package.json: No license field
"
`;

exports[`Run command should execute "package-one test" and exit with 0: stdout 1`] = `
"$ echo testing
testing
"
`;

exports[`Run command should execute "package-three start" and exit with 1: stderr 1`] = `
"Browserslist: caniuse-lite is outdated. Please run next command \`yarn upgrade\`
☔️ error No matching packages found
"
`;

exports[`Run command should execute "package-three start" and exit with 1: stdout 1`] = `""`;

exports[`Run command should execute "package-two start" and exit with 1: stderr 1`] = `
"Browserslist: caniuse-lite is outdated. Please run next command \`yarn upgrade\`
☔️ error an identifier must only match a single package but \\"package-two\\" matches the following packages: 
☔️ error @manypkg/basic-fixture-pkg-two
☔️ error @manypkg/basic-fixture-pkg-two-one
"
`;

exports[`Run command should execute "package-two start" and exit with 1: stdout 1`] = `""`;

exports[`Run command should execute "package-two-one something" and exit with 1: stderr 1`] = `
"Browserslist: caniuse-lite is outdated. Please run next command \`yarn upgrade\`
warning package.json: No license field
warning ../../package.json: No license field
error Command \\"something\\" not found.
"
`;

exports[`Run command should execute "package-two-one something" and exit with 1: stdout 1`] = `
"info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
"
`;

exports[`Run command should execute "package-two-one start" and exit with 0: stderr 1`] = `
"Browserslist: caniuse-lite is outdated. Please run next command \`yarn upgrade\`
warning package.json: No license field
warning ../../package.json: No license field
"
`;

exports[`Run command should execute "package-two-one start" and exit with 0: stdout 1`] = `
"$ echo hello start
hello start
"
`;

exports[`Run command should execute "packages/package-two start" and exit with 0: stderr 1`] = `
"Browserslist: caniuse-lite is outdated. Please run next command \`yarn upgrade\`
warning package.json: No license field
warning ../../package.json: No license field
"
`;

exports[`Run command should execute "packages/package-two start" and exit with 0: stdout 1`] = `
"$ echo hello start
hello start
"
`;

exports[`Run command should execute "pkg-one start" and exit with 0: stderr 1`] = `
"Browserslist: caniuse-lite is outdated. Please run next command \`yarn upgrade\`
warning package.json: No license field
warning ../../package.json: No license field
"
`;

exports[`Run command should execute "pkg-one start" and exit with 0: stdout 1`] = `
"$ echo hello start
hello start
"
`;

exports[`Run command should execute "pkg-two start" and exit with 1: stderr 1`] = `
"Browserslist: caniuse-lite is outdated. Please run next command \`yarn upgrade\`
☔️ error an identifier must only match a single package but \\"pkg-two\\" matches the following packages: 
☔️ error @manypkg/basic-fixture-pkg-two
☔️ error @manypkg/basic-fixture-pkg-two-one
"
`;

exports[`Run command should execute "pkg-two start" and exit with 1: stdout 1`] = `""`;

exports[`Run command should execute "pkg-two-one start" and exit with 0: stderr 1`] = `
"Browserslist: caniuse-lite is outdated. Please run next command \`yarn upgrade\`
warning package.json: No license field
warning ../../package.json: No license field
"
`;

exports[`Run command should execute "pkg-two-one start" and exit with 0: stdout 1`] = `
"$ echo hello start
hello start
"
`;
